plugin/evm/vm.go

```diff
@@ -674,38 +698,62 @@ func (vm *VM) initChainState(lastAcceptedBlock *types.Block) error {
 func (vm *VM) SetState(_ context.Context, state snow.State) error {
 	switch state {
 	case snow.StateSyncing:
-		vm.bootstrapped = false
+		vm.bootstrapped.Set(false)
 		return nil
 	case snow.Bootstrapping:
-		vm.bootstrapped = false
-		if err := vm.StateSyncClient.Error(); err != nil {
-			return err
-		}
-		// After starting bootstrapping, do not attempt to resume a previous state sync.
-		if err := vm.StateSyncClient.ClearOngoingSummary(); err != nil {
-			return err
-		}
-		// Ensure snapshots are initialized before bootstrapping (i.e., if state sync is skipped).
-		// Note calling this function has no effect if snapshots are already initialized.
-		vm.blockChain.InitializeSnapshots()
-		return nil
+		return vm.onBootstrapStarted()
 	case snow.NormalOp:
-		// Initialize goroutines related to block building once we enter normal operation as there is no need to handle mempool gossip before this point.
-		if err := vm.initBlockBuilding(); err != nil {
-			return fmt.Errorf("failed to initialize block building: %w", err)
-		}
-		vm.bootstrapped = true
-		return nil
+		return vm.onNormalOperationsStarted()
 	default:
 		return snow.ErrUnknownState
 	}
 }
 
-// initBlockBuilding starts goroutines to manage block building
-func (vm *VM) initBlockBuilding() error {
+// onBootstrapStarted marks this VM as bootstrapping
+func (vm *VM) onBootstrapStarted() error {
+	vm.bootstrapped.Set(false)
+	if err := vm.StateSyncClient.Error(); err != nil {
+		return err
+	}
+	// After starting bootstrapping, do not attempt to resume a previous state sync.
+	if err := vm.StateSyncClient.ClearOngoingSummary(); err != nil {
+		return err
+	}
+	// Ensure snapshots are initialized before bootstrapping (i.e., if state sync is skipped).
+	// Note calling this function has no effect if snapshots are already initialized.
+	vm.blockChain.InitializeSnapshots()
+
+	return nil
+}
+
+// onNormalOperationsStarted marks this VM as bootstrapped
+func (vm *VM) onNormalOperationsStarted() error {
+	if vm.bootstrapped.Get() {
+		return nil
+	}
+	vm.bootstrapped.Set(true)
+
 	ctx, cancel := context.WithCancel(context.TODO())
 	vm.cancel = cancel
 
+	// update validators first
+	if err := vm.performValidatorUpdate(ctx); err != nil {
+		return fmt.Errorf("failed to update validators: %w", err)
+	}
+	vdrIDs := vm.validatorState.GetNodeIDs().List()
+	// then start tracking with updated validators
+	if err := vm.uptimeManager.StartTracking(vdrIDs); err != nil {
```

Comment:
It looks like this is the only place where we call `StartTracking`. Do we also need to call `StartTracking` for new validators that are added during normal execution?

Reply:
Gotcha, thanks. The `StartTracking` name is a little confusing in that case, but it was pre-existing from the interface so not worth changing.

Could you add the context that `StartTracking` is used to update the uptimes of existing validators to account for the time we were offline as a comment on this line though?